src/firmware/guest/types/snp.rs

```diff
@@ -117,14 +128,300 @@ bitfield! {
 /// ReportedTcb value as the TCB version to derive the VCEK. This value is set by the hypervisor.
+/// The VLEK is generated externally and has to be loaded into the machine.
 ///
 /// The firmware guarantees that the ReportedTcb value is never greater than the installed TCB
 /// version
+///
+/// Since the release of the 1.56 ABI, the Attestation Report was bumped from version 2 to 3.
+/// Due to content differences, both versions are kept separately in order to provide backwards compatibility and most reliable security.
+#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
+pub enum AttestationReport {
+    /// Version 2 of the Attestation Report
+    V2(AttestationReportV2),
+    /// Version 3 of the Attestation Report
+    V3(AttestationReportV3),
+}
+
+impl TryFrom<&[u8]> for AttestationReport {
+    type Error = AttestationReportError;
+
+    fn try_from(raw_report: &[u8]) -> Result<Self, Self::Error> {
+        let version =
+            u32::from_le_bytes([raw_report[0], raw_report[1], raw_report[2], raw_report[3]]);
+        // Return the appropriate report version
+        match version {
+            2 => {
+                let report_v2: AttestationReportV2 = raw_report.try_into()?;
+                Ok(AttestationReport::V2(report_v2))
+            }
+            3 => {
+                let report_v3: AttestationReportV3 = raw_report.try_into()?;
+                Ok(AttestationReport::V3(report_v3))
+            }
+            _ => Err(AttestationReportError::UnsupportedReportVersion(version))?,
+        }
+    }
+}
+
+impl Attestable for AttestationReport {
+    fn measurable_bytes(&self) -> io::Result<Vec<u8>> {
+        //Return measurable bytes for the report
+        match self {
+            Self::V2(report) => {
+                let measurable_bytes: &[u8] = &bincode::serialize(report).map_err(|e| {
+                    Error::new(
+                        ErrorKind::Other,
+                        format!("Unable to serialize bytes: {}", e),
+                    )
+                })?;
+                Ok(measurable_bytes[..0x2a0].to_vec())
+            }
+            Self::V3(report) => {
+                let measurable_bytes: &[u8] = &bincode::serialize(report).map_err(|e| {
+                    Error::new(
+                        ErrorKind::Other,
+                        format!("Unable to serialize bytes: {}", e),
+                    )
+                })?;
+                Ok(measurable_bytes[..0x2a0].to_vec())
+            }
+        }
+    }
+    fn signature(&self) -> &Signature {
+        match self {
+            Self::V2(report) => &report.signature,
+            Self::V3(report) => &report.signature,
+        }
+    }
+}
+
+impl AttestationReport {
+    /// Get Attestation Report Version
+    pub fn version(&self) -> u32 {
+        match self {
+            Self::V2(report) => report.version,
+            Self::V3(report) => report.version,
```

Comment:
You can simplify this into a one-liner which saves typing.
```suggestion
            Self::V2(report) | Self::V3(report)  => report.version,
```

Reply:
It won't allow you to do this since both reports are not the same type. `Self::V2(report)` is of type `AttestationReportV2` and `Self::V3(report)` is of type `AttestationReportV3`

Reply:
https://www.rustexplorer.com/b#LyoKW2RlcGVuZGVuY2llc10KCiovCgpzdHJ1Y3QgQSB7dmVyc2lvbjogdTMyfQpzdHJ1Y3QgQiB7dmVyc2lvbjogdTMyfQoKZW51bSBMZXR0ZXJzIHsKICAgIEEoQSksCiAgICBCKEIpCn0KCmltcGwgTGV0dGVycyB7CiAgICBmbiB2ZXJzaW9uKCZzZWxmKSAtPiB1MzIgewogICAgICAgIG1hdGNoIHNlbGYgewogICAgICAgICAgICBTZWxmOjpBKGxldHRlcikgfCBTZWxmOjpCKGxldHRlcikgPT4gbGV0dGVyLnZlcnNpb24KICAgICAgICB9CiAgICB9Cn0KZm4gbWFpbigpIHsKICAgIHByaW50bG4hKCIiKTsKfQ==

Reply:
You are totally right. My mistake